src/net/data_stream.rs

```diff
@@ -73,216 +107,143 @@ pub struct AdcDacData {
 /// `stream` is the logically consumer (UDP transmitter) of the enqueued data.
 pub fn setup_streaming(
     stack: NetworkReference,
-) -> (BlockGenerator, DataStream) {
-    let queue = cortex_m::singleton!(: Queue<AdcDacData, BLOCK_BUFFER_SIZE> = Queue::new()).unwrap();
-
+) -> (FrameGenerator, DataStream) {
+    let queue =
+        cortex_m::singleton!(: Queue<StreamFrame, FRAME_COUNT> = Queue::new())
+            .unwrap();
     let (producer, consumer) = queue.split();
 
-    let generator = BlockGenerator::new(producer);
+    let frame_pool =
+        cortex_m::singleton!(: Pool<[u8; FRAME_SIZE]>= Pool::new()).unwrap();
+
+    // Note(unsafe): We guarantee that FRAME_DATA is only accessed once in this function.
+    let memory = unsafe { &mut FRAME_DATA };
+    frame_pool.grow(memory);
+
+    let generator = FrameGenerator::new(producer, frame_pool);
 
-    let stream = DataStream::new(stack, consumer);
+    let stream = DataStream::new(stack, consumer, frame_pool);
 
     (generator, stream)
 }
 
-/// The data generator for a stream.
-pub struct BlockGenerator {
-    queue: Producer<'static, AdcDacData, BLOCK_BUFFER_SIZE>,
-    current_id: u16,
+struct StreamFrame {
+    format: StreamFormat,
+    sequence_number: u16,
+    buffer: Box<[u8; FRAME_SIZE], Init>,
+    offset: usize,
+    batch_count: u16,
+    batch_size: u8,
```

Comment:
I think this is more complicated and heavy then necessary.
I would make the `FrameGenerator` generic (or by some other means knowledgable) over the (all compile-time constants) length of a frame, the number of batches per frame and the format. Also make the frame size an even multiple of the batch size (plus header). Full checking is easier that way and offset and batch count become irrelevant.
Then there is no need to put all this additional data (format, sequence_number, offset, batch_count, batch_size) into each item in the Queue. Remember this all gets copied. Everything apart from the frame sequence number is constant.
The sequence number together with the format should be added to the frame after the `alloc()` in `add()` (or just before `enqueue()`).

Also see the comment on `setup_streaming()`.

Reply:
The batch count and batch size are not for this driver, but rather for the UDP receiver. Is your comment more around just removing them from this struct and having them only exist in the header (since they're const generics, we don't have to carry them around, as they're part of the type signature)?

Reply:
Both.
The receiver knows batch size and content from batch format ID. It knows batches_per_frame either from the format ID (if we let the application define that) or simply by the size of the UDP packet (they are not split).
The other aspect is that `StreamFrame` is the queue item and is being copied into and out of it. None of the data apart from `buffer` should be in the queue because it is either known/constant or has no bearing on the processing at the other end of the queue.

Reply:
I'm going to elect to keep things not templated on batch size and batch count per packet. The template parameters start to get really messy and cause a lot of pain in other areas of the codebase. I think it would make more sense to make them just constant struct members. I'll get a small refactor going to clean things up generally in the aspects mentioned above though.